gloo/benchmark/runner.cc

```diff
@@ -187,37 +197,72 @@ void Runner::run(BenchmarkFn<T>& fn, int n) {
       barrier_->run();
     }
 
-    // Switch mode based on iteration count or time spent
-    auto iterations = options_.iterationCount;
-    if (iterations <= 0) {
-      GLOO_ENFORCE_GT(options_.iterationTimeNanos, 0);
+    benchmarks.push_back(std::move(benchmark));
+  }
 
-      Distribution warmup;
-      for (int i = 0; i < options_.warmupIterationCount; i++) {
-        Timer dt;
-        benchmark->run();
-        warmup.add(dt);
-      }
+  // Switch mode based on iteration count or time spent
+  auto iterations = options_.iterationCount;
+  if (iterations <= 0) {
+    GLOO_ENFORCE_GT(options_.iterationTimeNanos, 0);
+
+    // Create warmup jobs for every thread
+    std::vector<std::unique_ptr<RunnerJob>> jobs;
+    for (auto i = 0; i < options_.threads; i++) {
+      auto fn = [&benchmark = benchmarks[i]] { benchmark->run(); };
+      auto job = make_unique<RunnerJob>(fn, options_.warmupIterationCount);
+      jobs.push_back(std::move(job));
+    }
 
-      // Broadcast duration of fastest iteration during warmup,
-      // so all nodes agree on the number of iterations to run for.
-      auto nanos = broadcast(warmup.min());
-      iterations = std::max(1L, options_.iterationTimeNanos / nanos);
+    // Start jobs on every thread (synchronized across processes)
```

Comment:
It's fine if it's not super accurate. They will kick off at the same time, so the hope is they run in parallel and we get some good measurement of how long an iteration takes under the configured parallelism. Totally best effort.